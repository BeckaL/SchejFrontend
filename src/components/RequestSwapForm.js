import React, { Component } from 'react'
import axios from 'axios'
import Popup from "reactjs-popup";

export default class RequestSwapForm extends Component {
  constructor(props){
    super(props)
    this.state = {
      chosenShift: '',
      userShift: "",
      userShifts: []
    }
  }

  componentWillMount() {
    axios.get(
      `${process.env.REACT_APP_API_URL}/shiftsbyuser/${localStorage['id']}`
    )
    .then(response => {
      this.setState({
        userShifts: response.data.sort((shift, secondShift) => {
          return shift.start_time - secondShift.start_time
        }),
        chosenShift: this.props.shiftInfo.shiftId
      })
    })
  }

  handleSubmit = (event) => {
    event.preventDefault()
    console.log(this.state)
    axios.post(
      `${process.env.REACT_APP_API_URL}/requests`,
      {
        requested_shift_id: this.state.chosenShift,
        current_shift_id: this.state.userShift
      }
    )
    .then(response => {
      console.log(response)
    })
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
    setTimeout(() => {
      console.log(this.state)
    }, 100)
  }

  formatDate = (time) => {
    return new Date(parseInt(time)).toLocaleString('en-GB')
  }

  render() {
    return(
      <form id="request-swap-form" onSubmit={this.handleSubmit}>
        <span className='shift-info-label'>Select a shift:</span>
        <select className="menu" name="userShift" onChange={this.handleChange} value={this.state.userShift}>
          <option className='menu-item'>My shifts</option>
          {this.state.userShifts.map( (shift, index) => {
            if (new Date(parseInt(shift.start_time)) > new Date()) {
              return <option className="menu-item" key={index} value={shift.id}>{this.formatDate(shift.start_time)}-{this.formatDate(shift.end_time)}</option>
            }
          })}
        </select>
        <br />
        <button id='confirm-swap-request-form' className='custom-button'>Confirm</button>
      </form>
    )
  }
}
