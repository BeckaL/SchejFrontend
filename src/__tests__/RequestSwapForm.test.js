import React from 'react'
import axios from 'axios'
import { mount, shallow } from 'enzyme'
import RequestSwapForm from '../components/RequestSwapForm'
import MockDate from 'mockdate'

jest.mock('axios')
jest.useFakeTimers()

describe('Request swap form', () => {
  var wrapper

  beforeAll(() => {
    global.localStorage['id'] = 1
    MockDate.set(1545151751163)

    const response = {
      data: [
        {
          start_time: new Date().getTime(),
          end_time: new Date().getTime()
        },
        {
          start_time: new Date().getTime(),
          end_time: new Date().getTime()
        }]
    }
    axios.get.mockResolvedValue(response)

    wrapper = shallow(<RequestSwapForm shiftInfo={{ shiftId: 1 }}/>)
  })

  test('it matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('it renders the correct form', () => {
    expect(wrapper.find('#request-swap-form').length).toEqual(1)
  })

  test('it renders a select box', () => {
    expect(wrapper.find('select').length).toEqual(1)
  })

  test('it renders a button to confirm a swap request', () => {
    const button = wrapper.find('#confirm-swap-request-form')
    expect(button.text()).toEqual('Confirm')
  })

  test('it makes the correct request to the API', () => {
    const select = wrapper.find('select')
    const form = wrapper.find('#request-swap-form')

    const response = {}

    axios.post.mockResolvedValue(response)
    select.simulate('change', { target: { name: 'userShift', value: 2 } })
    form.simulate('submit', { preventDefault() {} })

    jest.runAllTimers()
    expect(axios.post).toHaveBeenCalledTimes(1)
    expect(axios.post).toHaveBeenCalledWith(
      `http://localhost:3001/api/v1/requests`, {
        requested_shift_id: 1,
        current_shift_id: 2
      }
    )
  })
})
