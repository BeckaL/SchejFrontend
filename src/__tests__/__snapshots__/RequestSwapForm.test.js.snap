// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request swap form it matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RequestSwapForm
    shiftInfo={
      Object {
        "shiftId": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="shift-info-label"
        >
          Select a shift:
        </span>,
        <select
          className="menu"
          name="userShift"
          onChange={[Function]}
          value=""
        >
          <option
            className="menu-item"
          >
            My shifts
          </option>
          <option
            className="menu-item"
          >
            18 Dec, 16:49
             - 
            18 Dec, 16:49
          </option>
          <option
            className="menu-item"
          >
            18 Dec, 16:49
             - 
            18 Dec, 16:49
          </option>
        </select>,
        <br />,
        <button
          className="custom-button"
          id="confirm-swap-request-form"
        >
          Confirm
        </button>,
      ],
      "id": "request-swap-form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Select a shift:",
          "className": "shift-info-label",
        },
        "ref": null,
        "rendered": "Select a shift:",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              className="menu-item"
            >
              My shifts
            </option>,
            Array [
              <option
                className="menu-item"
              >
                18 Dec, 16:49
                 - 
                18 Dec, 16:49
              </option>,
              <option
                className="menu-item"
              >
                18 Dec, 16:49
                 - 
                18 Dec, 16:49
              </option>,
            ],
          ],
          "className": "menu",
          "name": "userShift",
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "My shifts",
              "className": "menu-item",
            },
            "ref": null,
            "rendered": "My shifts",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "18 Dec, 16:49",
                " - ",
                "18 Dec, 16:49",
              ],
              "className": "menu-item",
              "value": undefined,
            },
            "ref": null,
            "rendered": Array [
              "18 Dec, 16:49",
              " - ",
              "18 Dec, 16:49",
            ],
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "18 Dec, 16:49",
                " - ",
                "18 Dec, 16:49",
              ],
              "className": "menu-item",
              "value": undefined,
            },
            "ref": null,
            "rendered": Array [
              "18 Dec, 16:49",
              " - ",
              "18 Dec, 16:49",
            ],
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Confirm",
          "className": "custom-button",
          "id": "confirm-swap-request-form",
        },
        "ref": null,
        "rendered": "Confirm",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="shift-info-label"
          >
            Select a shift:
          </span>,
          <select
            className="menu"
            name="userShift"
            onChange={[Function]}
            value=""
          >
            <option
              className="menu-item"
            >
              My shifts
            </option>
            <option
              className="menu-item"
            >
              18 Dec, 16:49
               - 
              18 Dec, 16:49
            </option>
            <option
              className="menu-item"
            >
              18 Dec, 16:49
               - 
              18 Dec, 16:49
            </option>
          </select>,
          <br />,
          <button
            className="custom-button"
            id="confirm-swap-request-form"
          >
            Confirm
          </button>,
        ],
        "id": "request-swap-form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Select a shift:",
            "className": "shift-info-label",
          },
          "ref": null,
          "rendered": "Select a shift:",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                className="menu-item"
              >
                My shifts
              </option>,
              Array [
                <option
                  className="menu-item"
                >
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </option>,
                <option
                  className="menu-item"
                >
                  18 Dec, 16:49
                   - 
                  18 Dec, 16:49
                </option>,
              ],
            ],
            "className": "menu",
            "name": "userShift",
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "My shifts",
                "className": "menu-item",
              },
              "ref": null,
              "rendered": "My shifts",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "18 Dec, 16:49",
                  " - ",
                  "18 Dec, 16:49",
                ],
                "className": "menu-item",
                "value": undefined,
              },
              "ref": null,
              "rendered": Array [
                "18 Dec, 16:49",
                " - ",
                "18 Dec, 16:49",
              ],
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "18 Dec, 16:49",
                  " - ",
                  "18 Dec, 16:49",
                ],
                "className": "menu-item",
                "value": undefined,
              },
              "ref": null,
              "rendered": Array [
                "18 Dec, 16:49",
                " - ",
                "18 Dec, 16:49",
              ],
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Confirm",
            "className": "custom-button",
            "id": "confirm-swap-request-form",
          },
          "ref": null,
          "rendered": "Confirm",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
